### PR TITLE
Added tree_height.cpp in dynamic_programming directory

### PR DESCRIPTION
**Pull request by mann2108**
#### Description of Change

Added finding the height of the valid tree in O(n) using bottom - up dp approach.
where n is a number of nodes.
A valid tree contain exactly n-1 edges when n is number of nodes.
Sample Input: - 
4
1 2
1 3
2 3

Hence the height of the tree is 3

#### Checklist

- [X] Added description of change
- [X] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTION.md#New-File-Name-guidelines)
- [X] Added tests and example, test must pass
- [X] Relevant documentation/comments is changed or added
- [X] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTION.md#Commit-Guidelines)
- [X] Search previous suggestions before making a new one, as yours may be a duplicate.
- [ ] Sort by alphabetical order
- [X] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: dynamic_programming/tree_height.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
+/// C++ Program to find height of the tree using bottom - up DP.
```
Comment (cclauss) [inline dynamic_programming/tree_height.cpp:1]: Why /// instead of // ??

**Comment by cclauss** [comment]
Four errors found in https://github.com/TheAlgorithms/C-Plus-Plus/actions

File: dynamic_programming/tree_height.cpp (hunk 1 above)
Comment (mann2108) [reply dynamic_programming/tree_height.cpp:1]: Noted.
Actually in my IDE it gives better commenting visuals compare to //.

Comment (cclauss) [reply dynamic_programming/tree_height.cpp:1]: What IDE do you use?

Comment (mann2108) [reply dynamic_programming/tree_height.cpp]: I prefer to use CodeBlocks

Comment (mann2108) [reply dynamic_programming/tree_height.cpp]: Capture_1
Capture_2

**Comment by mann2108** [comment]
> Four errors found in https://github.com/TheAlgorithms/C-Plus-Plus/actions

@cclauss Its fixed.

File: dynamic_programming/tree_height.cpp [hunk 2]
```diff
@@ -29,31 +29,31 @@ void dp_with_dfs(int u) {
         if (!visited[v]) {
             dp_with_dfs(v);
 
-            /// selected maximum subtree height from all childs.
+            // selected maximum subtree height from all child.
```
Comment (cclauss) [inline dynamic_programming/tree_height.cpp:32]: children
```suggestion
            // select maximum subtree height from all children.
```

Comment (bhaumikmistry) [reply dynamic_programming/tree_height.cpp:32]: Good catch

**Comment by bhaumikmistry** [comment]
@cclauss Any idea what can we do to add tests for incoming checks, we can tests the working for this by just looking at it!!

**Comment by cclauss** [comment]
There are multiple [unit test frameworks](https://docs.microsoft.com/en-us/visualstudio/test/writing-unit-tests-for-c-cpp?view=vs-2019) for C++.  So someone who really knows the current state of C++ (i.e. not me) needs to pick one of them and then I can help with the GitHub Actions to enforce the rule that new submissions must have tests and pass them.  If you want to proceed, please open a new issue.

**Comment by mann2108** [comment]
@cclauss Should we merge this PR?

**Comment by cclauss** [comment]
It has my approval but let's let @bhaumikmistry decide.

File: dynamic_programming/tree_height.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+// global declarations
+// no of nodes max limit.
+const int MAX = 1e5;
+// adjacency list
+std::vector<int> adj[MAX];
+std::vector<bool> visited;
+std::vector<int> dp;
+
+void dp_with_dfs(int u) {
```
Comment (cclauss) [inline dynamic_programming/tree_height.cpp:25]:
```suggestion
void depth_first_search(int u) {
```

File: dynamic_programming/tree_height.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    for (int v : adj[u]) {
+        if (!visited[v]) {
+            dp_with_dfs(v);
+
+            // select maximum sub-tree height from all children.
+            child_height = std::max(child_height, dp[v]+1);
+        }
+    }
+    // assigned the max child height to current visited node.
+    dp[u] = child_height;
```
Comment (cclauss) [inline dynamic_programming/tree_height.cpp:37]: What is a `dp` and why do I have to guess?

File: dynamic_programming/tree_height.cpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+int main() {
+    // number of nodes
+    int n;
+    std::cin >> n;
+    int u, v;
```
Comment (cclauss) [inline dynamic_programming/tree_height.cpp:44]: Can't we come up with more self-documenting variable names?  Single character variable names are so old school.

File: dynamic_programming/tree_height.cpp [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        }
+    }
+    // assigned the max child height to current visited node.
+    dp[u] = child_height;
+}
+
+int main() {
+    // number of nodes
+    int n;
+    std::cin >> n;
```
Comment (cclauss) [inline dynamic_programming/tree_height.cpp:43]: Why not prompt the user to tell them what the program is looking for?

File: dynamic_programming/tree_height.cpp (hunk 4 above)
Comment (mann2108) [reply dynamic_programming/tree_height.cpp:37]: The general way to write a dp equation is termed as dp. Which means "Dynamic Programming".

File: dynamic_programming/tree_height.cpp (hunk 5 above)
Comment (mann2108) [reply dynamic_programming/tree_height.cpp:44]: Will update it.

File: dynamic_programming/tree_height.cpp (hunk 6 above)
Comment (mann2108) [reply dynamic_programming/tree_height.cpp:43]: Yeah will update it.

File: dynamic_programming/tree_height.cpp (hunk 5 above)
Comment (poyea) [reply dynamic_programming/tree_height.cpp:46]: @mann2108 Can you resolve this also?

Comment (mann2108) [reply dynamic_programming/tree_height.cpp:46]: Yeah I resolve n as no_of_nodes.

Comment (poyea) [reply dynamic_programming/tree_height.cpp:54]: Ok, I see.

File: dynamic_programming/tree_height.cpp [hunk 7]
```diff
@@ -8,7 +8,14 @@
...
  * 1 3
  * 2 4
- * Height of the tree : - 3
+ * which can be represented as                 
+ *   1                      
+ *  / \                     
+ * 2   3                    
+ * |                        
+ * 4            
+ *
```
Comment (mann2108) [inline dynamic_programming/tree_height.cpp:17]: @poyea Yeah good idea to show the tree like this.

**Comment by poyea** [comment]
LGTM. Can be merged at any time.

**Pull request merged**